tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php

```diff
@@ -555,49 +556,313 @@ public function testTrimExpressionInvalidMode() : void
-        self::assertEquals('2009-04-01', date('Y-m-d', strtotime($row['sub_quarters'])), 'Subtracting quarters should end up on 2009-10-01');
-        self::assertEquals('2016-01-01', date('Y-m-d', strtotime($row['add_years'])), 'Adding years should end up on 2016-01-01');
-        self::assertEquals('2004-01-01', date('Y-m-d', strtotime($row['sub_years'])), 'Subtracting years should end up on 2004-01-01');
+    /**
+     * @dataProvider modeProvider
+     */
+    public function testDateSubQuarters(callable $buildQuery, callable $bindParams) : void
+    {
+        $this->assertDateExpression(
+            $buildQuery,
+            $bindParams,
+            static function (AbstractPlatform $platform, string $interval) : string {
+                return $platform->getDateSubQuartersExpression('test_datetime', $interval);
+            },
+            3,
+            '2009-04-01 10:10:10'
+        );
+    }
+
+    /**
+     * @dataProvider modeProvider
+     */
+    public function testDateAddYears(callable $buildQuery, callable $bindParams) : void
+    {
+        $this->assertDateExpression(
+            $buildQuery,
+            $bindParams,
+            static function (AbstractPlatform $platform, string $interval) : string {
+                return $platform->getDateAddYearsExpression('test_datetime', $interval);
+            },
+            6,
+            '2016-01-01 10:10:10'
+        );
+    }
+
+    /**
+     * @dataProvider modeProvider
+     */
+    public function testDateSubYears(callable $buildQuery, callable $bindParams) : void
+    {
+        $this->assertDateExpression(
+            $buildQuery,
+            $bindParams,
+            static function (AbstractPlatform $platform, string $interval) : string {
+                return $platform->getDateSubYearsExpression('test_datetime', $interval);
+            },
+            6,
+            '2004-01-01 10:10:10'
+        );
+    }
+
+    /**
+     * @param callable $buildQuery Builds the portion of the query representing the interval value
+     * @param callable $bindParams Binds the interval value to the statement
+     * @param callable $expression Builds the platform-specific interval expression
+     * @param int      $interval   Interval value
+     * @param string   $expected   Expected value
+     */
+    private function assertDateExpression(callable $buildQuery, callable $bindParams, callable $expression, int $interval, string $expected) : void
+    {
+        $connection = $this->connection;
+        $platform   = $connection->getDatabasePlatform();
+
+        $query = sprintf('SELECT %s FROM fetch_table', $expression($platform, $buildQuery($interval)));
+        $stmt  = $connection->prepare($query);
+        $bindParams($stmt, $interval);
+
+        $stmt->execute();
+
+        $date = $stmt->fetchColumn();
+
+        $this->assertEquals($expected, date('Y-m-d H:i:s', strtotime($date)));
+    }
+
+    /**
+     * @return mixed[][]
```

Comment:
I used `mixed[][]` deliberately, same as for all test data providers in the codebase.

The fact that the provider returns all arguments as a 2nd level array doesn't make that array list of a certain type since all arguments are independent. Having it declared as `mixed[][]` opens up the possibility to add new returned arguments without having to convert the type to something meaningless like `callable[][]|int[][]`.

Reply:
Works for me - we'll see if more analysis tools start complaining about this one later on tho.

Merging then :+1: